File: dt241m-controller/internal/discovery/discovery.go
package discovery

import (
	"context"
	"sync"
	"time"

	"github.com/jacobgad/dt241m-controller/internal/dt241m"
)

type Hit struct {
	IP   string
	Info *dt241m.DeviceInfo
}

type Options struct {
	Concurrency  int
	Timeout      time.Duration
	OnHit        func(Hit)
	OnProbeStart func(ip string, active int)
}

// Probe queries every address with bounded concurrency; unreachable addresses are simply skipped.
func Probe(ctx context.Context, client dt241m.Client, ips []string, opts Options) []Hit {
	var (
		mu     sync.Mutex
		hits   []Hit
		active int
		wg     sync.WaitGroup
		queue  = make(chan string)
	)
	workers := opts.Concurrency
	if workers < 1 {
		workers = 1
	}
	if workers > len(ips) {
		workers = len(ips)
	}
	for i := 0; i < workers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for ip := range queue {
				if ctx.Err() != nil {
					continue
				}
				mu.Lock()
				active++
				if opts.OnProbeStart != nil {
					opts.OnProbeStart(ip, active)
				}
				mu.Unlock()
				probeCtx, cancel := context.WithTimeout(ctx, opts.Timeout)
				info, err := client.GetDeviceInfo(probeCtx, ip)
				cancel()
				if err == nil {
					hit := Hit{IP: ip, Info: info}
					mu.Lock()
					hits = append(hits, hit)
					mu.Unlock()
					if opts.OnHit != nil {
						opts.OnHit(hit)
					}
				}
				mu.Lock()
				active--
				mu.Unlock()
			}
		}()
	}
	for _, ip := range ips {
		if ctx.Err() != nil {
			break
		}
		queue <- ip
	}
	close(queue)
	wg.Wait()
	return hits
}

File: dt241m-controller/src/discovery.ts
import { mapWithConcurrency } from "./concurrency.js";
import type { Dt241mClient } from "./dt241m/client.js";
import type { DeviceInfo } from "./dt241m/schemas.js";

export type ProbeHit = { ip: string; info: DeviceInfo };

export type ProbeOptions = {
  concurrency: number;
  timeoutMs: number;
  onHit?: (hit: ProbeHit) => void | Promise<void>;
  onProbeStart?: (ip: string, activeCount: number) => void;
  shouldStop?: () => boolean;
};

export async function probeAddresses(client: Dt241mClient, ips: readonly string[], options: ProbeOptions): Promise<ProbeHit[]> {
  const hits: ProbeHit[] = [];
  let active = 0;

  await mapWithConcurrency(ips, options.concurrency, async (ip) => {
    if (options.shouldStop?.()) return;
    active += 1;
    options.onProbeStart?.(ip, active);
    try {
      const info = await client.getDeviceInfo(ip, { timeoutMs: options.timeoutMs });
      const hit = { ip, info };
      hits.push(hit);
      await options.onHit?.(hit);
    } catch {
      // why: an unreachable address is the normal case during a subnet sweep, not a scan failure
    } finally {
      active -= 1;
    }
  });

  return hits;
}
